File: backend/controllers/budgetController.js
const mongoose = require('mongoose');
const Budget = require('../models/Budget');

// @desc    Create a new budget
// @route   POST /api/budgets
// @access  Private
exports.createBudget = async (req, res) => {
  try {
    const { categoryId, limitAmount, month, year } = req.body;
    const userId = req.user.userId;

    if (!categoryId || !limitAmount || !month || !year) {
      return res.status(400).json({ message: 'All fields are required' });
    }

    // Check if a budget already exists for this category in the given month/year
    const existingBudget = await Budget.findOne({
      userId,
      categoryId,
      month,
      year,
    });

    if (existingBudget) {
      return res.status(400).json({ message: 'Budget for this category already exists for the given month.' });
    }

    const budget = new Budget({
      userId,
      categoryId,
      limitAmount,
      month,
      year,
    });

    await budget.save();
    res.status(201).json(budget);
  } catch (error) {
    console.error('Error creating budget:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Update a budget
// @route   PUT /api/budgets/:id
// @access  Private
exports.updateBudget = async (req, res) => {
  try {
    const { id } = req.params;
    const { limitAmount } = req.body;
    const userId = req.user.userId;

    const budget = await Budget.findOne({ _id: id, userId });
    if (!budget) {
      return res.status(404).json({ message: 'Budget not found' });
    }

    budget.limitAmount = limitAmount || budget.limitAmount;
    await budget.save();

    res.json(budget);
  } catch (error) {
    console.error('Error updating budget:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

// @desc    Delete a budget
// @route   DELETE /api/budgets/:id
// @access  Private
exports.deleteBudget = async (req, res) => {
  try {
    const { id } = req.params;
    const userId = req.user.userId;

    const budget = await Budget.findOne({ _id: id, userId });
    if (!budget) {
      return res.status(404).json({ message: 'Budget not found' });
    }

    await budget.deleteOne();
    res.json({ message: 'Budget removed' });
  } catch (error) {
    console.error('Error deleting budget:', error);
    res.status(500).json({ message: 'Server error' });
  }
};

File: backend/src/main/java/com/personalfinance/backend/controllers/BudgetController.java
package com.personalfinance.backend.controllers;

import com.personalfinance.backend.models.Budget;
import com.personalfinance.backend.repositories.BudgetRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/budgets")
public class BudgetController {

    @Autowired
    private BudgetRepository budgetRepository;

    // API: Thiết lập (tạo mới) một ngân sách (Budget) cho một danh mục chi tiêu
    @PostMapping
    public ResponseEntity<?> createBudget(@RequestBody Budget request) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        
        List<Budget> existing = budgetRepository.findByUserIdAndMonthAndYear(userId, request.getMonth(), request.getYear());
        if (existing.stream().anyMatch(b -> b.getCategoryId().equals(request.getCategoryId()))) {
             return ResponseEntity.badRequest().body(Map.of("message", "Budget for this category already exists for the given month."));
        }

        request.setUserId(userId);
        Budget saved = budgetRepository.save(request);
        return ResponseEntity.status(201).body(saved);
    }

    // API: Cập nhật thông tin ngân sách (ví dụ: thay đổi hạn mức chi tiêu)
    @PutMapping("/{id}")
    public ResponseEntity<?> updateBudget(@PathVariable String id, @RequestBody Map<String, Object> updates) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        Optional<Budget> bOpt = budgetRepository.findById(id);
        
        if (bOpt.isEmpty() || !bOpt.get().getUserId().equals(userId)) {
            return ResponseEntity.status(404).body(Map.of("message", "Budget not found"));
        }
        
        Budget b = bOpt.get();
        if (updates.containsKey("limitAmount")) {
            b.setLimitAmount(Double.parseDouble(updates.get("limitAmount").toString()));
        }
        budgetRepository.save(b);
        return ResponseEntity.ok(b);
    }

    // API: Xóa một thiết lập ngân sách khỏi hệ thống
    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteBudget(@PathVariable String id) {
        String userId = SecurityContextHolder.getContext().getAuthentication().getName();
        Optional<Budget> bOpt = budgetRepository.findById(id);
        
        if (bOpt.isEmpty() || !bOpt.get().getUserId().equals(userId)) {
            return ResponseEntity.status(404).body(Map.of("message", "Budget not found"));
        }
        
        budgetRepository.deleteById(id);
        return ResponseEntity.ok(Map.of("message", "Budget removed"));
    }
}
